### PR TITLE
Enforcing the fee structure

### PR DESCRIPTION
**Pull request by CurbShifter**
A proposal for a quick check to enforce the minimum fee for the amount of slots used in a block relative to the number of transactions. 

This check implies that if someone 'overpays' for a slot, another tx could still have a lower fee while the block stays valid. This added block validation condition improves fair distribution of fees among the blocks and miners. 

Note that this change makes fat block forging still possible when really needed. By sending a tx with a high fee.

### CODE REVIEW

**Comment by CurbShifter** [comment]
Added another option with a stricter fee verification, that doesn't allow paying for other transactions. (And rejects fat blocks). Because that feature would still favor rich miners. Where the miner would use a privately held UT (not propagated) which is only self rewarding if a block is forged.

File: src/brs/BlockchainProcessorImpl.java [hunk 1]
```diff
@@ -975,13 +980,32 @@ private void pushBlock(final Block block) throws BlockNotAcceptedException {
...
           digest.update(transaction.getBytes());
         }
 
         if (calculatedTotalAmount > block.getTotalAmountNQT()
             || calculatedTotalFee > block.getTotalFeeNQT()) {
           throw new BlockNotAcceptedException("Total amount or fee don't match transaction totals for block " + block.getHeight());
         }
+
+		if (Burst.getFluxCapacitor().isActive(FeatureToggle.SLOT_FEE_ENFORCING_OPTION_2)) {
+			Arrays.sort(feeArray);
```
Comment (blankey1337) [inline src/brs/BlockchainProcessorImpl.java]: Do you mind switching this to use spaces vs tabs for consistency?

File: src/brs/BlockchainProcessorImpl.java [hunk 2]
```diff
@@ -975,13 +980,32 @@ private void pushBlock(final Block block) throws BlockNotAcceptedException {
...
+			for (int slotIdx = 0; slotIdx < feeArray.length; slotIdx++) {
+				if (feeArray[slotIdx] >= Constants.FEE_QUANT * (slotIdx + 1)) {
+					throw new BlockNotAcceptedException("Transaction fee is not enough to be included in this block " + block.getHeight());
+				}   
+			}
+		}
+		
+        if (Burst.getFluxCapacitor().isActive(FeatureToggle.SLOT_FEE_ENFORCING)
+			&& Constants.FEE_QUANT * ((block.getTransactions().size() / 2) * (1 + block.getTransactions().size())) > block.getTotalFeeNQT())) {
```
Comment (blankey1337) [inline src/brs/BlockchainProcessorImpl.java]: This probably deserves to be broken out into it's own function and properly unit tested.

File: src/brs/BlockchainProcessorImpl.java [hunk 3]
```diff
@@ -975,13 +980,32 @@ private void pushBlock(final Block block) throws BlockNotAcceptedException {
...
         }
 
         if (calculatedTotalAmount > block.getTotalAmountNQT()
             || calculatedTotalFee > block.getTotalFeeNQT()) {
           throw new BlockNotAcceptedException("Total amount or fee don't match transaction totals for block " + block.getHeight());
         }
+
+		if (Burst.getFluxCapacitor().isActive(FeatureToggle.SLOT_FEE_ENFORCING_OPTION_2)) {
+			Arrays.sort(feeArray);
+			for (int slotIdx = 0; slotIdx < feeArray.length; slotIdx++) {
```
Comment (blankey1337) [inline src/brs/BlockchainProcessorImpl.java]: You seem to be overloading `slotIdx` here which makes this code hard to follow.  Maybe rename `slotIdx` on line `985` to something else?

**Comment by CurbShifter** [comment]
yes ofc. committed your suggestions. The unit test sounds like a good plan for the formula. Perhaps it needs to cast to a float for a proper division.

**Comment by rico666** [comment]
This is a hard fork change and as such should be added to some other code fork of the BRS.
